SOURCES/etc/xapi.d/plugins/smartctl.py

```diff
@@ -10,31 +10,30 @@
 from xcpngutils.operationlocker import OperationLocker
 
 @error_wrapped
-def _list_disks():
-    disks = []
+def _list_devices():
+    devices = []
     result = run_command(['smartctl', '--scan'])
     for line in result['stdout'].splitlines():
-        if line.startswith('/dev/') and not line.startswith('/dev/bus/'):
-            disks.append(line.split()[0])
-    return disks
+        devices.append({'name': line.split()[0], 'type':line.split()[2]})
+    return devices
 
 @error_wrapped
 def get_information(session, args):
     results = {}
     with OperationLocker():
-        disks = _list_disks()
-        for disk in disks:
-            cmd = run_command(["smartctl", "-j", "-a", disk], check=False)
-            results[disk] = json.loads(cmd['stdout'])
+        devices = _list_devices()
+        for device in devices:
+            cmd = run_command(["smartctl", "-j", "-a", "-d", device['name'], device['type']], check=False)
+            results[device] = json.loads(cmd['stdout'])
         return json.dumps(results)
 
 @error_wrapped
 def get_health(session, args):
     results = {}
     with OperationLocker():
-        disks = _list_disks()
-        for disk in disks:
-            cmd = run_command(["smartctl", "-j", "-H", disk])
+        devices = _list_devices()
+        for device in devices:
+            cmd = run_command(["smartctl", "-j", "-H", "-d", device['name'], device['type']], check=False)
             json_output = json.loads(cmd['stdout'])
             if json_output['smart_status']['passed']:
                 results[disk] = "PASSED"
```

Comment:
`disk` looks wrong... should be device["name"] probably ;)

Reply:
I will test that immediatly

Reply:
That doesn't seems to work.

Reply:
Do you think that you can give a try using this one [smartctl.py](https://github.com/xcp-ng/xcp-ng-xapi-plugins/blob/gtn-pr-46/SOURCES/etc/xapi.d/plugins/smartctl.py) ?

Reply:
sure

Reply:
it's doesn't seems to work

Reply:
damn... can you try to run this on your host `xe host-call-plugin host-uuid=<HOSTUUID>  plugin=smartctl.py fn=health` and give me the output?
If needed you can have the host uuid using `xe host-list`

Reply:
![image](https://github.com/user-attachments/assets/be881e9a-a340-4107-84c0-26f49a7e23cb)

Reply:
I think I found the issue. This new version of [smartctl.py](https://github.com/xcp-ng/xcp-ng-xapi-plugins/blob/gtn-pr-46/SOURCES/etc/xapi.d/plugins/smartctl.py) should work better. Can you give it a try?
And thanks a lot for your help and patience for testing it :+1: I don't have such hardware myself so it is really appreciate.

Reply:
So the last version should work :crossed_fingers:

Reply:
@1234Erwan , Have you had the opportunity to test the latest version?

Reply:
Hi @gthvn1, I pulled your latest smartctl.py and gave it a spin. It runs successfully on my host (XO reports `All disks are healthy`; no errors in `/var/log/xensource.log`)

For context:

```
# smartctl --scan
/dev/sda -d scsi # /dev/sda, SCSI device
/dev/sdb -d scsi # /dev/sdb, SCSI device
/dev/sdc -d scsi # /dev/sdc, SCSI device
/dev/sdd -d scsi # /dev/sdd, SCSI device
/dev/bus/0 -d megaraid,4 # /dev/bus/0 [megaraid_disk_04], SCSI device
/dev/bus/0 -d megaraid,5 # /dev/bus/0 [megaraid_disk_05], SCSI device
/dev/bus/0 -d megaraid,6 # /dev/bus/0 [megaraid_disk_06], SCSI device
/dev/bus/0 -d megaraid,7 # /dev/bus/0 [megaraid_disk_07], SCSI device
```